cryoet_data_portal_neuroglancer/precompute/contrast_limits.py

```diff
@@ -0,0 +1,557 @@
+        peak_gradient = gradient[largest_peak]
+        # Find the start gradient percentage
+        # Before the gradient climbs above 20% of the peak gradient
+        # Find the median values of the gradient
+        start_of_rising = np.where(gradient > gradient_threshold * peak_gradient)[0][0]
+        mean_before_rising = np.mean(gradient[:start_of_rising])
+        start_gradient_threshold = mean_before_rising / peak_gradient
+
+        # Find the end gradient percentage
+        end_of_flattening = np.where(gradient[start_of_rising:] < gradient_threshold * peak_gradient)[0][0]
+        mean_after_rising = np.median(gradient[start_of_rising + end_of_flattening :])
+        end_gradient_threshold = mean_after_rising / peak_gradient
+
+        start_gradient_threshold = max(0.01, start_gradient_threshold)
+        end_gradient_threshold = max(0.01, end_gradient_threshold)
+
+        return start_gradient_threshold, end_gradient_threshold
+
+    def _caculate_cdf(self, n_bins):
+        min_value = np.min(self.volume.flatten())
+        max_value = np.max(self.volume.flatten())
+        hist, bin_edges = np.histogram(self.volume.flatten(), bins=n_bins, range=[min_value, max_value])
+        cdf = np.cumsum(hist) / np.sum(hist)
+        try:
+            gradient = np.gradient(cdf.compute())
+        except AttributeError:
+            gradient = np.gradient(cdf)
+        x = np.linspace(min_value, max_value, n_bins)
+        return cdf, bin_edges, gradient, x
+
+    def compute_contrast_limit(
+        self,
+        gradient_threshold: float = 0.3,
+    ) -> tuple[float, float]:
+        """Calculate the contrast limits using the Cumulative Distribution Function.
+
+        Parameters
+        ----------
+        gradient_threshold: float, optional.
+            The threshold multiplier against the peak gradient.
+            This is used to estimate the start and end of the contrast limits.
+            By default 0.3.
+
+        Returns
+        -------
+            tuple[float, float]
+                The calculated contrast limits.
+        """
+        # Calculate the histogram of the volume
+        cdf, bin_edges, gradient, x = self._caculate_cdf(n_bins=512)
+
+        # Find the largest peak in the gradient
+        largest_peak = np.argmax(gradient)
+        peak_gradient_value = gradient[largest_peak]
+
+        # Find the start and end gradient percentages
+        start_gradient, end_gradient = self.automatic_parameter_estimation(gradient_threshold)
+
+        # Find where the gradient starts rising and starts flattening after the peak
+        start_of_rising = np.where(gradient > start_gradient * peak_gradient_value)[0][0]
+        end_of_flattening = np.where(gradient[largest_peak:] < end_gradient * peak_gradient_value)[0][0]
+        end_of_flattening += largest_peak
+        start_limit = bin_edges[start_of_rising]
+        end_limit = bin_edges[end_of_flattening]
+
+        self.cdf = [x, cdf]
```

Comment:
Perhaps it could be interesting to remove it, and in the manual testing module, to make an explicit call to `_calculate_cdf` to get the cdf and the `x` range instead.

Reply:
sounds good, i'll make it a public function and call it in the manual test